presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveCrossJoinWithConstantInput.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.matching.Captures;
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.spi.plan.JoinType;
+import com.facebook.presto.spi.plan.PlanNode;
+import com.facebook.presto.spi.plan.ProjectNode;
+import com.facebook.presto.spi.plan.ValuesNode;
+import com.facebook.presto.spi.relation.ConstantExpression;
+import com.facebook.presto.spi.relation.RowExpression;
+import com.facebook.presto.spi.relation.VariableReferenceExpression;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.IntStream;
+
+import static com.facebook.presto.SystemSessionProperties.isRemoveCrossJoinWithConstantSingleRowInputEnabled;
+import static com.facebook.presto.sql.planner.PlannerUtils.addProjections;
+import static com.facebook.presto.sql.planner.plan.Patterns.join;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+
+/**
+ * When one side of a cross join is one single row of constant, we can remove the cross join and replace it with a project.
+ * <pre>
+ * - Cross Join
+ *      - table scan
+ *          leftkey
+ *      - values // only one row
+ *          rightkey := 1
+ * </pre>
+ * into
+ * <pre>
+ * - project
+ *      leftkey := leftkey
+ *      rightkey := 1
+ *      - table scan
+ *          leftkey
+ * </pre>
+ */
+public class RemoveCrossJoinWithConstantInput
+        implements Rule<JoinNode>
+{
+    @Override
+    public Pattern<JoinNode> getPattern()
+    {
+        return join().matching(x -> x.getType().equals(JoinType.INNER) && x.getCriteria().isEmpty());
```

Comment:
What about filte in the join node? WE should push it down as well?

Reply:
Good point. I remember that the filter field in join node is only populated for outer join node, but not for inner join node. But still this case needs to be handled in this rule.
I updated the code to create a filter node, and this filter node will be pushed down with push filter down rules later.

Reply:
@kaikalur can you please mark this comment as resolved if your original comment as addressed.